src/Hint/NewType.hs

```diff
@@ -65,3 +82,78 @@ hasNoStrategy _                      = False
 withNewtype :: Deriving a -> Deriving a
 withNewtype (Deriving l Nothing rs)  = Deriving l (Just $ DerivNewtype l) rs
 withNewtype d                        = d
+
+data WarnNewtype = WarnNewtype
+    { newDecl :: Hs.LHsDecl Hs.GhcPs
+    , insideType :: Hs.HsType Hs.GhcPs
+    }
+
+-- | Given a declaration, returns the suggested \"newtype\"ized declaration following these guidelines:
+-- * Types ending in a \"#\" are __ignored__, because they are usually unboxed primitives - @data X = X Int#@
+-- * @ExistentialQuantification@ stuff is __ignored__ - @data X = forall t. X t@
+-- * Constructors with (nonempty) constraints are __ignored__ - @data X a = (Eq a) => X a@
+-- * Single field constructors get newtyped - @data X = X Int@ -> @newtype X = X Int@
+-- * Single record field constructors get newtyped - @data X = X {getX :: Int}@ -> @newtype X = X {getX :: Int}@
+-- * All other declarations are ignored.
+singleSimpleFieldNew :: Hs.LHsDecl Hs.GhcPs -> Maybe WarnNewtype
+singleSimpleFieldNew (Hs.L loc (Hs.TyClD ext decl@(Hs.DataDecl _ name _ _ def@(Hs.HsDataDefn _ Hs.DataType ctx _ _ [Hs.L _ constructor] derives))))
+    | Just inType <- simpleCons constructor =
+        Just WarnNewtype
+              { newDecl = Hs.L loc $ Hs.TyClD ext decl {Hs.tcdDataDefn = def
+                  { Hs.dd_ND = Hs.NewType
+                  , Hs.dd_cons = map (\(Hs.L consloc x) -> Hs.L consloc $ dropConsBang x) $ Hs.dd_cons def
+                  }}
+              , insideType = inType
+              }
+singleSimpleFieldNew _ = Nothing
+
+-- TODO: check for MAGIC#HASH
+-- TODO: get tests to pass
+--
+-- | Checks whether its argument is a \"simple constructor\" (see criteria in 'singleSimpleFieldNew')
+-- returning the type inside the constructor if it is. This is needed for bang/MagicHash analysis.
+simpleCons :: Hs.ConDecl Hs.GhcPs -> Maybe (Hs.HsType Hs.GhcPs)
+simpleCons (Hs.ConDeclH98 _ _ _ [] context (Hs.PrefixCon [Hs.L _ inType]) _)
+    | emptyOrNoContext context
+    , not $ isUnboxedTuple inType
+    , not $ isHashy inType
+    = Just inType
+simpleCons (Hs.ConDeclH98 _ _ _ [] context (Hs.RecCon (Hs.L _ [Hs.L _ (Hs.ConDeclField _ [_] (Hs.L _ inType) _)])) _)
+    | emptyOrNoContext context
+    , not $ isUnboxedTuple inType
+    , not $ isHashy inType
+    = Just inType
+simpleCons _ = Nothing
+
+isHashy :: Hs.HsType Hs.GhcPs -> Bool
```

Comment:
Probably not the most reliable way to print a type in general, but it *should* work here, because the `#` is always at the end (?)

Reply:
Also, `isKindHash` from `Utils` will be unused after this, I will delete it.